fix LoopLimiter bug(702)

XiaoyuHe opened the pull request.

ltfish commented:
Can you please describe in the description what bug you are trying to fix? Thanks!

XiaoyuHe commented:
LoopLimiter fails after angr merges path's code to sim_state:
Fail code fragment:
    pg = proj.factory.path_group(p)
    tech = angr.exploration_techniques.LoopLimiter(count = 3)
    pg.use_technique(tech)
    pg.explore()
Fail tested program:
#include<stdio.h>
int main(){
   while(1){
      printf("hellowolrd\n");
   }
   return 0;
}
fix description :
   1. Remove code in LoopLimiter which rescues the path removed by LoopLimiter. 
   2.  add two kinds of detections:
       (1) simple loop that original angr's loop detection had:
              while(1){
                  a =b +c;
              }
      (2)loop with function that original angr's loop detection didn't have. 
              while(1){
                  haha("hello");//haha is a function
              }

ltfish commented on angr/sim_state.py:
```diff
@@ -436,6 +436,20 @@ def step(self, **kwargs):
+    def detect_loops(self, n=None):
+        mc1 = self.callstack.top.block_counter.most_common()
+        mc2 = collections.Counter([l for l in self.history.descriptions]).most_common()
+        if len(mc1) == 0 and len(mc2) ==0 :
```
`len(mc2) == 0:`

ltfish commented on angr/sim_state.py:
```diff
@@ -436,6 +436,20 @@ def step(self, **kwargs):
+            return None
+        elif len(mc2) == 0 and len(mc1) > 0:
+            return mc1[0][1]
+        elif len(mc2) >0  and len(mc1) ==0:
```
- `len(mc2) > 0`
- `len(mc1) == 0`

ltfish commented on angr/sim_state.py:
```diff
@@ -436,6 +436,20 @@ def step(self, **kwargs):
+            return mc1[0][1]
+        elif len(mc2) >0  and len(mc1) ==0:
+            return mc2[0][1]
+        elif mc1[0][1]>mc2[0][1]:
```
Spaces around the `>` please.

ltfish commented on angr/exploration_techniques/looplimiter.py:
```diff
@@ -16,7 +16,9 @@ def __init__(self, count=5, discard_stash='spinning'):
+        '''
         if len(pg.stashes[stash]) == 0 and len(pg.stashes[self.discard_stash]) > 0:
             pg.stashes[stash].append(pg.stashes[self.discard_stash].pop())
+        '''
```
Just remove this piece of code please. We use Git for version control.

ltfish commented on angr/sim_state.py:
```diff
@@ -436,6 +436,20 @@ def step(self, **kwargs):
 
+    def detect_loops(self, n=None):
+        mc1 = self.callstack.top.block_counter.most_common()
+        mc2 = collections.Counter([l for l in self.history.descriptions]).most_common()
```
Directly accessing `self.history.descriptions` makes me feel sad. `history.descriptions` is a collection of string representations of `SimSuccessors`. They are meant to be read and used by humans. You should consider using `history.bbl_addrs` instead. At least, you avoid performing many unnecessary string comparisons.

ltfish commented:
Linting failed: https://travis-ci.org/angr/angr/jobs/294424468

It'll be great if you can provide a test case with this PR :-)

ltfish commented on angr/sim_state.py:
```diff
@@ -436,6 +436,20 @@ def step(self, **kwargs):
         """
         return self.project.factory.successors(self, **kwargs)
 
+    def detect_loops(self, n=None):
```
It seems that `n` is not used.

ltfish commented:
Also, are you aware of [this PR](https://github.com/angr/angr/pull/712)?

XiaoyuHe commented:
@ltfish . Thanks a lot for your help. I have changed the code and added a test case. All of them passed the tests. Please check it again.  tyb0807's PR needs to recovery the CFG, it's easy to be broken for huge program and time consuming. Maybe my commit is a choice. It also has some drawbacks in some cases.

ltfish commented:
> I have changed the code and added a test case. All of them passed the tests.

Thanks! I appreciate that.

> Maybe my commit is a choice.

It is indeed a choice.

> It also has some drawbacks in some cases.

Can you please elaborate?

ltfish commented on tests/test_looplimiter.py:
```diff
@@ -0,0 +1,15 @@
+    pg.use_technique(tech)
+    pg.explore()
+    nose.tools.assert_equal(len(pg.stashes['spinning']), 1)
+if __name__ == '__main__':
```
Sorry for being picky. Two blank lines before `if __name__ == '__main__':` please.

XiaoyuHe commented:
@ltfish for example:
Fail code fragment:
```
    pg = proj.factory.path_group(p)
    tech = angr.exploration_techniques.LoopLimiter(count = 3)
    pg.use_technique(tech)
    pg.explore()
```
Fail tested program:
#include<stdio.h>
```
int main(){
    printf("helloworld\n");
    printf("helloworld\n");
    printf("helloworld\n");
    printf("helloworld\n");
    printf("helloworld\n");
    return 0;
}
```

XiaoyuHe commented:
@ltfish ,it is passed.
